README.md
# PythonProject

## Description
The project is called Scientific Python Interactive Data Acoustic Modeling, or SPIDAM. 
It reads an audio file and adds the data into a graph chosen by the user.
It creates a GUI that the user can use to upload the audio file and to visualize the data.

## Basic Instructions.
Run the program and wait for the GUI to initialize.
Click on the 'Upload audio' button.
Choose the clap audio file.
Open the dropdown menu and select the graph you want.

## Why does it exist?
This is a group project for school.
It allows us to gain a better understanding with Python and its libraries. 
It teaches us on how to work with a team.
=======

CHANGELOG.md
- Added basic comments to file and added requirements (by hoxtibals at Thu Nov 23 18:15:00 2023 -0500)
- Created python modules (by hoxtibals at Thu Nov 23 12:19:56 2023 -0500)
- Updated changelog as well as script (by hoxtibals at Thu Nov 23 12:00:29 2023 -0500)
- Merge branch 'main' of https://github.com/hoxtibals/PythonProject (by hoxtibals at Thu Nov 23 11:56:44 2023 -0500)
- This is a test for the testing.txt (by hoxtibals at Thu Nov 23 11:56:26 2023 -0500)
- Updated changelog Script (by hoxtibals at Thu Nov 23 11:53:38 2023 -0500)
- This is a test commit (by hoxtibals at Mon Nov 6 12:08:45 2023 -0500)
- Create update_changelog.sh (by hoxtibals at Mon Nov 6 12:00:16 2023 -0500)
- This is the first commit (by hoxtibals at Mon Nov 6 11:45:30 2023 -0500)
- Initial commit (by hoxtibals at Mon Nov 6 11:27:51 2023 -0500)
- created open WAVfile function (by hoxtibals at Thu Nov 30 16:32:16 2023 -0500)
- Added more view labels (by hoxtibals at Thu Nov 30 12:59:18 2023 -0500)
- Created button in view w/ basic load file command (by hoxtibals at Mon Nov 27 12:33:11 2023 -0500)
- Added basic comment to file and added requirements (by hoxtibals at Thu Nov 23 18:15:00 2023 -0500)
- Created python modules (by hoxtibals at Thu Nov 23 12:19:56 2023 -0500)
- Updated changelog as well as script (by hoxtibals at Thu Nov 23 12:00:29 2023 -0500)
- Merge branch 'main' of https://github.com/hoxtibals/PythonProject (by hoxtibals at Thu Nov 23 11:56:44 2023 -0500)
- This is a test for the testing.txt (by hoxtibals at Thu Nov 23 11:56:26 2023 -0500)
- Updated changelog Script (by hoxtibals at Thu Nov 23 11:53:38 2023 -0500)
- This is a test commit (by hoxtibals at Mon Nov 6 12:08:45 2023 -0500)
- Create update_changelog.sh (by hoxtibals at Mon Nov 6 12:00:16 2023 -0500)
- This is the first commit (by hoxtibals at Mon Nov 6 11:45:30 2023 -0500)
- Initial commit (by hoxtibals at Mon Nov 6 11:27:51 2023 -0500)
- set up the basic's for all files (by hoxtibals at Thu Nov 30 22:27:40 2023 -0500)
- Added setter for controllers and create app module (by hoxtibals at Thu Nov 30 16:54:46 2023 -0500)
- created open WAVfile function (by hoxtibals at Thu Nov 30 16:32:16 2023 -0500)
- Added more view labels (by hoxtibals at Thu Nov 30 12:59:18 2023 -0500)
- Created button in view w/ basic load file command (by hoxtibals at Mon Nov 27 12:33:11 2023 -0500)
- Added basic comment to file and added requirements (by hoxtibals at Thu Nov 23 18:15:00 2023 -0500)
- Created python modules (by hoxtibals at Thu Nov 23 12:19:56 2023 -0500)
- Updated changelog as well as script (by hoxtibals at Thu Nov 23 12:00:29 2023 -0500)
- Merge branch 'main' of https://github.com/hoxtibals/PythonProject (by hoxtibals at Thu Nov 23 11:56:44 2023 -0500)
- This is a test for the testing.txt (by hoxtibals at Thu Nov 23 11:56:26 2023 -0500)
- Updated changelog Script (by hoxtibals at Thu Nov 23 11:53:38 2023 -0500)
- This is a test commit (by hoxtibals at Mon Nov 6 12:08:45 2023 -0500)
- Create update_changelog.sh (by hoxtibals at Mon Nov 6 12:00:16 2023 -0500)
- This is the first commit (by hoxtibals at Mon Nov 6 11:45:30 2023 -0500)
- Initial commit (by hoxtibals at Mon Nov 6 11:27:51 2023 -0500)
- re arranged some code (by hoxtibals at Tue Dec 5 11:00:52 2023 -0500)
- added hide_message method (by hoxtibals at Tue Dec 5 10:45:46 2023 -0500)
- Added more exception handling (by hoxtibals at Tue Dec 5 10:40:50 2023 -0500)
- created basic load stats button (by hoxtibals at Tue Dec 5 10:27:46 2023 -0500)
- added basic summary stats to window (by hoxtibals at Mon Dec 4 16:54:37 2023 -0500)
- updated changelog (by hoxtibals at Mon Dec 4 12:22:17 2023 -0500)
- Basic Program that can open a WAV file (by hoxtibals at Mon Dec 4 12:11:35 2023 -0500)
- changed attributes for model (by hoxtibals at Mon Dec 4 12:09:09 2023 -0500)
- updated more of the main application (by hoxtibals at Mon Dec 4 11:42:36 2023 -0500)
- updated git ignore (by hoxtibals at Mon Dec 4 11:41:55 2023 -0500)
- Created basic gitignore file for pycache (by hoxtibals at Mon Dec 4 11:37:10 2023 -0500)
- finished calculating reverb (by hoxtibals at Mon Dec 4 11:25:15 2023 -0500)
- started calculate reverb function (by hoxtibals at Mon Dec 4 10:48:23 2023 -0500)
- basic contribution for setting freq (by hoxtibals at Mon Dec 4 10:33:09 2023 -0500)
- changed some attributes (by hoxtibals at Mon Dec 4 10:27:35 2023 -0500)
- created basic target frequency function (by hoxtibals at Sun Dec 3 22:54:31 2023 -0500)
- updated attributes with get/set methods (by hoxtibals at Sun Dec 3 22:27:35 2023 -0500)
- added basic error handling&parsing between modules (by hoxtibals at Sun Dec 3 12:08:06 2023 -0500)
- added basic stats computation for channels & lngth (by hoxtibals at Sun Dec 3 11:52:50 2023 -0500)
- created basic attributes in model and updated view (by hoxtibals at Sun Dec 3 03:18:27 2023 -0500)
- Merge branch 'main' of https://github.com/hoxtibals/PythonProject (by hoxtibals at Sat Dec 2 01:33:18 2023 -0500)
- Python interpreter failure (by hoxtibals at Sat Dec 2 01:33:14 2023 -0500)
- Merge branch 'main' of https://github.com/hoxtibals/PythonProject (by hoxtibals at Fri Dec 1 12:44:37 2023 -0500)
- Syncing PC's (by hoxtibals at Fri Dec 1 12:44:32 2023 -0500)
- deleted test (by hoxtibals at Thu Nov 30 23:37:49 2023 -0500)
- set up the basic's for all files (by hoxtibals at Thu Nov 30 22:27:40 2023 -0500)
- Added setter for controllers and create app module (by hoxtibals at Thu Nov 30 16:54:46 2023 -0500)
- created open WAVfile function (by hoxtibals at Thu Nov 30 16:32:16 2023 -0500)
- Added more view labels (by hoxtibals at Thu Nov 30 12:59:18 2023 -0500)
- Created button in view w/ basic load file command (by hoxtibals at Mon Nov 27 12:33:11 2023 -0500)
- Added basic comment to file and added requirements (by hoxtibals at Thu Nov 23 18:15:00 2023 -0500)
- Created python modules (by hoxtibals at Thu Nov 23 12:19:56 2023 -0500)

model.py
#It consists of pure application logic, which interacts with the database. 
# It includes all the information to represent data to the end user.
# this is where we would handle any WAV file manipulation

#Create a WAV file class that can be used to manipulate the WAV file and read metadata  
# the controller module will be used to call these methods in model module 

from scipy.signal import welch
import numpy as np
import matplotlib.pyplot as matplt
from matplotlib.figure import Figure
from pydub import AudioSegment
import os
import subprocess
import io
from scipy.fft import fft



class Model:
    def __init__(self):
        '''
        attributes of the WAV file we will be using
        '''
        self._sample_rate = 0
        self._data = 0
        self._num_channels = 0
        self._length = 0
        self._avgRT = 0
        self._spectrum = np.empty((0,0))
        self._freqs = np.array([])
        self._t = np.array([])
        self._graphs = {}
        self._resFreq = 0

    @property
    def graphs(self):
        return self._graphs
    @property
    def length(self):
        return self._length
    @property
    def num_channels(self):
        return self._num_channels
    @property
    def sample_rate(self):
        return self._sample_rate
    @property
    def data(self):
        return self._data
    @property
    def spectrum(self):
        return self._spectrum
    @property
    def freqs(self):
        return self._freqs
    @property
    def t(self):
        return self._t
    @property
    def avgRT(self):
        return self._avgRT
    @property
    def resFreq(self):
        return self._resFreq
    @sample_rate.setter
    def sample_rate(self,value):
        self._sample_rate = value
    @spectrum.setter
    def spectrum(self, value):
        self._spectrum = value
    @freqs.setter
    def freqs(self, value):
        self._freqs = value
    @t.setter
    def t(self, value):
        self._t = value
    @num_channels.setter
    def num_channels(self,value):
        self._num_channels = value
    @length.setter
    def length(self,value):
        self._length = value   
    @graphs.setter
    def graphs(self,name,graph):
        self._graphs[name] = graph 
    @avgRT.setter
    def avgRT(self,value):
        self._avgRT = round(abs(value),2)
    @resFreq.setter
    def resFreq(self,value):
        self._resFreq = round(value,2)
    @data.setter    
    def data(self,value):
        # we can get number of channels by looking at the data
        self._data = value
        self._num_channels = self.data.shape[len(self.data.shape)-1]
        self._length = self.data.shape[0]/self.sample_rate
    
    '''
    input: a filepath to a file
    output: update attributes with WAV file object
    '''
    def openWAVfile(self,filepath):
        #open the WAV file and convert to a WAV file
        # now we load the WAV file but first we gotta handle multiple channels
        try:
            new_path = self.convert_to_wav(filepath)
            audio = AudioSegment.from_wav(new_path)
            self.length = round(audio.duration_seconds,2)
            self.num_channels = audio.channels
            mono_audio = audio.set_channels(1)
            # Convert mono_audio to numpy array
            self._data = np.array(mono_audio.get_array_of_samples())
            self._sample_rate = mono_audio.frame_rate
            self.spectrum, self.freqs, self.t, im = matplt.specgram(self.data, Fs=self.sample_rate, NFFT=1024, cmap=matplt.get_cmap("jet"))
            self.avgRT = (self.calculate_reverb(1000) + self.calculate_reverb(250) + self.calculate_reverb(6000))/3
            self.findRes()
        except FileNotFoundError:
            raise FileNotFoundError("File is not a WAV file")
        except ValueError:
            raise ValueError("File is NOT a Audio file")
        
    '''
    just strip the metadata from the file so we can read it correctly
    '''
    def strip_metadata(self,input_file, output_file):
        command = ['ffmpeg', '-i', input_file, '-map_metadata', '-1', '-c:v', 'copy', '-c:a', 'copy', output_file]
        subprocess.run(command, check=True)
        
    '''
    input: a filepath to a file
    output: a WAV file object which will not have metadata 
    '''
    def convert_to_wav(self,filepath):
        # List of known audio file extensions
        
        audioFormats = {
        'wav': None,
        'mp3': AudioSegment.from_mp3,
        'ogg': AudioSegment.from_ogg,
        'flv': AudioSegment.from_flv,
        'aac': lambda fp: AudioSegment.from_file(fp, format='aac'),
        'wma': lambda fp: AudioSegment.from_file(fp, format='wma'),
        'aiff': lambda fp: AudioSegment.from_file(fp, format='aiff'),
        'flac': lambda fp: AudioSegment.from_file(fp, format='flac'),
        'alac': lambda fp: AudioSegment.from_file(fp, format='alac'),
        'm4a': lambda fp: AudioSegment.from_file(fp, format='m4a'),
    }
        # Get the file extension
        file_extension = filepath.split('.')[-1].lower()

        # If the file is not an audio file, raise an exception
        if file_extension not in audioFormats:
            raise ValueError(f"File '{filepath}' is not a recognized audio file.")

        # If the file is not a WAV file, convert it to WAV format
        if file_extension != 'wav':
            # Get the appropriate AudioSegment method
            method = audioFormats[file_extension]
            # Open the file with the appropriate method
            audio = method(filepath)
            # Export the audio in WAV format to a BytesIO object
            wav_file = io.BytesIO()
            audio.export(wav_file, format='wav')
            wav_file.seek(0)  # Go back to the start of the file

            return wav_file

            # Return the path to the new file
            """else:
            return filepath"""
        # If the file is already a WAV file, strip metadata and return the path
        else:
            stripped_wav_filepath = os.path.splitext(filepath)[0] + '_stripped.wav'
            self.strip_metadata(filepath, stripped_wav_filepath)
            return stripped_wav_filepath
    '''
    input: a frequency in hz to be selected
    output: a target frequency around the 1000 hz range
    '''
    def target_freq(self,target):
        for x in self.freqs:
            if (x > target).any():
                break
        return x
    '''
    resonant frequency finder which does FFT to calculate the resonant frequency 
    '''
    def findRes(self):
        freq,power = welch(self.data,self.sample_rate,nperseg=4096)
        topFreq = freq[np.argmax(power)]
        self.resFreq = topFreq

    '''
    input:
    output: a dictionary of all the graphs we will need assigned to the object
    We just create the graphs here and assign them to the object in a dictionary
    '''
    def graph_figures(self):
        #create all figures we will need and assign to dictionary and return

        #create the figure for the spectogram
        Spectogram = Figure(figsize=(6,6),dpi=100)
        self.graphs['Spectogram'] = Spectogram
        #add the plot to the figure
        plot1 = Spectogram.add_subplot(111)
        #add the data to the plot
        _,_,_,im = plot1.specgram(self._data, Fs=self._sample_rate, NFFT=1024, cmap='jet')
        plot1.set_xlabel('Time(s)')
        plot1.set_ylabel('Frequency(Hz)')
        Spectogram.colorbar(im,ax=plot1).set_label('Intensity(dB)')
        #Spectogram.colorbar(im).set_label('Intensity(dB)')

        #Create Low freq figure
        rt60Low = round(abs(self.calculate_reverb(250)),2)
        lowFreq = Figure(figsize=(6,6),dpi=100)
        self.graphs['Low Frequency'] = lowFreq
        lowFreq.suptitle('Low Frequency Figure')
        lowFreq.text(0.95, 0.95, f'rt60: {rt60Low}seconds', horizontalalignment='right', verticalalignment='top', transform=lowFreq.transFigure)
        plot2 = lowFreq.add_subplot(111)
        plot2.plot(self.t,self.frequency_check(250), linewidth=0.5)
        plot2.set_xlabel('Time(s)')
        plot2.set_ylabel('Decibels(dB)')

        #create Mid freq figure
        rt60Mid = round(abs(self.calculate_reverb(1000)),2)
        midFreq = Figure(figsize=(6,6),dpi=100)
        self.graphs['Mid Frequency'] = midFreq
        midFreq.suptitle('High Frequency Figure')
        midFreq.text(0.95, 0.95, f'rt60: {rt60Mid} seconds', horizontalalignment='right', verticalalignment='top', transform=lowFreq.transFigure)
        plot3 = midFreq.add_subplot(111)
        plot3.plot(self.t,self.frequency_check(1000), linewidth=0.5)
        plot3.set_xlabel('Time(s)')
        plot3.set_ylabel('Decibels(dB)')

        # create High freq figure
        rt60High = round(abs(self.calculate_reverb(6000)),2)
        highFreq = Figure(figsize=(6,6),dpi=100)
        self.graphs['High Frequency'] = highFreq
        highFreq.suptitle('High Frequency Figure')
        highFreq.text(0.95, 0.95, f'rt60: {rt60High}seconds', horizontalalignment='right', verticalalignment='top', transform=lowFreq.transFigure)
        plot4 = highFreq.add_subplot(111)
        plot4.plot(self.t,self.frequency_check(4000), linewidth=0.5)
        plot4.set_xlabel('Time(s)')
        plot4.set_ylabel('Decibels(dB)')

        #extra plot 1, RAW amplitude from file
        amplitude = Figure(figsize=(6,6),dpi=100)
        self.graphs['Amplitude'] = amplitude
        amplitude.suptitle('RAW Amplitude Figure')
        plot5 = amplitude.add_subplot(111)
        time = np.arange(len(self.data))/self.sample_rate
        plot5.plot(time,abs(self.data), linewidth=0.5)
        plot5.set_xlabel('Time')
        plot5.set_ylabel('Amplitude')
        plot5.set_ylim(0, 10000)

        # Create a new figure for the combined frequency plot
        combinedFreq = Figure(figsize=(6,6),dpi=100)
        self.graphs['Combined Frequency'] = combinedFreq
        combinedFreq.suptitle('Combined Frequency Figure')
        plot6 = combinedFreq.add_subplot(111)
        # Plot the low, mid, and high frequency data on the same plot
        plot6.plot(self.t, self.frequency_check(250), label='Low Frequency', linewidth=0.5)
        plot6.plot(self.t, self.frequency_check(1000), label='Mid Frequency', linewidth=0.5)
        plot6.plot(self.t, self.frequency_check(4000), label='High Frequency', linewidth=0.5)
        plot6.set_xlabel('Time(s)')
        plot6.set_ylabel('Decibels(dB)')
        plot6.legend()

        # Create a new figure for the magnitude spectrum
        magnitudeSpectrum = Figure(figsize=(6,6),dpi=100)
        self.graphs['Magnitude Spectrum'] = magnitudeSpectrum
        magnitudeSpectrum.suptitle('Magnitude Spectrum Figure')
        plot7 = magnitudeSpectrum.add_subplot(111)
        # Compute and plot the magnitude spectrum
        Pxx, freqs, line = matplt.magnitude_spectrum(self._data, Fs=self._sample_rate, scale='dB', sides='default')
        plot7.plot(freqs, Pxx)
        plot7.set_xlabel('Frequency (Hz)')
        plot7.set_ylabel('Magnitude (dB)')

        #FFT graph
        fftGraph = Figure(figsize=(6,6),dpi=100)
        self.graphs['FFT'] = fftGraph
        fftGraph.suptitle('FFT Figure')
        plot8 = fftGraph.add_subplot(111)
        newData = np.abs(fft(self.data))
        plot8.plot(newData)
        plot8.set_xlabel('Frequency (Hz)')
        plot8.set_ylabel('Magnitude (dB)')

        #Freq vs Power
        freqPower = Figure(figsize=(6,6),dpi=100)
        self.graphs['Freq vs Power'] = freqPower
        freqPower.suptitle('Freq vs Power Figure')
        #here we will be using code given from the annoucement to add an extra graph
        freq,power = welch(self.data,self.sample_rate,nperseg=4096)
        n = len(self.data)
        k = np.arange(n)
        T = n/self.sample_rate
        freq = k/T
        freq = freq[:len(freq)//2]
        Y = np.fft.fft(self.data)/n
        Y = Y[:n//2]
        plot9 = freqPower.add_subplot(111)
        plot9.plot(freq,abs(Y))
        plot9.set_xscale('symlog')
        plot9.set_ylabel('Frequency (Hz)')
        plot9.set_xlabel('Power')


        
    
    '''
    input: a chosen frequency we are checking and turnign into decibles
    output:return the data of the frequency in decible which we 
    can pass directly to the plot directory
    '''
    def frequency_check(self,chosen_freq):
       #debugger(f'frequencies {self.freqs[:10]}')
       #1000 hz is the target frequency for mid range
        target_freq = self.target_freq(chosen_freq)
        #debugger(f'target frequency {target_freq}')
        index_freq = np.where(self.freqs == target_freq)[0][0]
        #debugger(f'index of target frequency {index_freq}')

        freq_data = self.spectrum[index_freq]
        #debugger(f'spectrum data {freq_data}')
        #add a very small constant to avoid 0
        decible_data = 10 * np.log10(freq_data + 1e-10)
        return decible_data
    '''
    input: a chosen frequency we are checking the reverb time for
    output: the reverb time for the frequency
    '''
    def calculate_reverb(self, chosen_freq):
        decible_data = self.frequency_check(chosen_freq)
        max_index = np.argmax(decible_data)
        value_max = decible_data[max_index]
        #debugger(f'max value {value_max}')
        #debugger(f'max index {max_index}')
        
        #this will get the rest of the array from Max to the end
        spliced_array = decible_data[max_index:]
        #now we can calculate tr20 and explogate it to rt60
        try:
            value_5db = self.nearestValue(spliced_array,value_max-5)
            index_5db = np.where(spliced_array == value_5db)[0][0]
            value_25db = self.nearestValue(spliced_array,value_max-25)
            index_25db = np.where(spliced_array == value_25db)[0][0]
            #now we have our rt20 which we can use to find rt60
            rt20 = self.t[index_5db] - self.t[index_25db]
            rt60 = rt20 * 3
        except IndexError:
            rt60 = 0
            raise IndexError("Frequency is not in the WAV file")
        finally:
            return rt60
    
    '''
    input: an array and a value
    output: the nearest value in the array to the passed value
    '''    
    def nearestValue(self,array,value):
        nparray = np.asarray(array)
        idx = (np.abs(nparray-value)).argmin()
        return nparray[idx]
        
        

    # have methods on what we will do to the WAV file and call in the controller module
def debugger(message):
    print(message)

controller.py
#It acts as an intermediary between view and model.
#It listens to the events triggered by view and queries model for the same.
#This is the file we will run to start the application 
# This will be where we will input the WAV file 
# View -> Controller -> Model



#Where we will call the show error view method
from view import View
from model import Model
import tkinter as tk

class Controller:
    def __init__(self,model,view):
        self.view = view
        self.model = model

    '''
    get the statistics from the model and pass them as a dictionary to the view
    '''
    def passStats(self):
        return {
            'sample_rate': self.model.sample_rate,
            'data': self.model.data,
            'num_channels': self.model.num_channels,
            'length': self.model.length,
            'spectrum': self.model.spectrum,
            'freqs': self.model.freqs,
            't': self.model.t,
            'average_rt60': self.model.avgRT,
            'resFreq': self.model.resFreq
        }
    '''
    handle the stats button click which will create the statistics
    '''
    def StatsButtonClicked(self):
        try:
            if not self.model.data:
                raise ValueError("No WAV file loaded")
            self.view.create_stats()
        except ValueError as error:
            self.view.show_error(error)   

    '''
    we add all possible graphs to the view and call them as needed from there
    '''
    def graphButtonClicked(self,frame):
        self.model.graph_figures()
        try:
            if not self.model.data.any():
                raise ValueError("No WAV file loaded")
            # we need to add all three graphs to the view figures attribute
            #toAdd = self.model.all_graphs()
            for figure in self.model.graphs:
                self.view.add_figure(figure,self.model.graphs[figure])
            self.view.displayGraph(self.model.graphs[self.view.graphDefault.get()],frame)
        except ValueError as error:
            self.view.show_error(error)

    '''
    can be used to set frequency from view (possible future feature)
    '''
    def setFrequency(self,freq):
        # call the method in the model to get the frequency 
        # call the method in the view to display the frequency
        self.model.target_freq(freq)

    '''
    The load wav file function which shows if we successfully load the file or not
    '''
    def loadWAVfile(self,filepath):
        # call the method in the model to load the WAV file
        # call the method in the view to display the WAV file
        try:
            self.model.openWAVfile(filepath)
            self.view.show_success("WAV file loaded successfully")
            #now we need to parse the info from opening the WAV file to the view
            #we can access our basic statistics from the model with the 
            #attributes for the model class
            self.view.create_stats()
        except ValueError as error:
            self.view.show_error(error)
        except FileNotFoundError as error:
            self.view.show_error(error)
        except Exception as error:  # Catch all other exceptions
            self.view.show_error("An unexpected error occurred: " + str(error))




view.py

#It represents the model’s data to user
import tkinter as tk
from tkinter import filedialog
import matplotlib.pyplot as matplt
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg

class View(tk.Frame):
    def __init__(self, parent):
        super().__init__(parent)
        '''
        TAKEN FROM CLASS 11-29 PARTTICIPATION
        we can use the this class to call the methods in the final module
        '''
        
        self.figures = {}
        # create widgets
        self.message_label = tk.Label(self, text='', foreground='black')
        self.message_label.pack()
        
        # save button
        self.WAVbutton = tk.Button(self, text='Open WAV file', command=self.loadWAVfileButton)
        self.WAVbutton.pack()
        
        #statistics window
        self.stats_label = tk.Label(self, text='Summary Stats: ', foreground='black')
        self.stats_button = tk.Button(self, text='Show Stats', command=self.showStatsButton)
        self.stats_label.pack()

        self.graphDefault = tk.StringVar()
        self.graphDefault.set('select graph')
        self.optionDropdown = tk.OptionMenu(parent, self.graphDefault, 'No figures')
        self.optionDropdown.pack()
        
        self.loadGraphButton = tk.Button(self, text='Show Graph', command=self.showGraphButton)
        
        

        # set the controller
        self.controller = None

    def add_figure(self, name, figure):
        """
        Add a figure to the dictionary of figures
        :param name: name of the figure
        :param figure: the figure
        :return:
        """
        self.figures[name] = figure

    def set_controller(self, controller):
        """
        Set the controller
        :param controller: the controller for the view to reference
        :return:
        """
        self.controller = controller
    
    def loadWAVfileButton(self):
        '''
        handle button click
        :return:
        '''
        if self.controller:
            path_file = filedialog.askopenfilename()
            self.controller.loadWAVfile(path_file)
            self.WAVbutton = tk.Button(self, text='Load Spectogram', command= self.showGraphButton)
            self.WAVbutton.pack()
        
    def set_freq_button(self, freq):
        '''
        Set the frequency
        :param freq:
        :return:
        '''
        #This will be called when button is pushed, will call controller to use command
        self.controller.setFrequency(freq)
        '''
        create the labels to show the statistics
        and get the statistics from the controller as a dictionary
        '''
    def create_stats(self):
        #we use controller methods to get the statistics from the model
        self.statisticsFrame = tk.Frame(self)
        self.statisticsFrame.pack()

        stats = self.controller.passStats()
        self.sampleRateLabel = tk.Label(self.statisticsFrame, text=f'Sample Rate: {stats["sample_rate"]}Hz')
        self.sampleRateLabel.pack()
        self.numChannelsLabel = tk.Label(self.statisticsFrame, text=f'Number of Channels (before change): {stats["num_channels"]}')
        self.numChannelsLabel.pack()
        self.lengthLabel = tk.Label(self.statisticsFrame, text=f'Length: {stats["length"]} seconds')
        self.lengthLabel.pack()
        self.average_rt60Label = tk.Label(self.statisticsFrame, text=f'Average RT60: {stats["average_rt60"]} seconds')
        self.average_rt60Label.pack()
        self.averageDifferentceLabel = tk.Label(self.statisticsFrame, text=f'Difference rt60 from 0.5: {stats["average_rt60"] - 0.5}seconds')
        self.averageDifferentceLabel.pack()
        self.resFreqLabel = tk.Label(self.statisticsFrame, text=f'Resonant Frequency: {stats["resFreq"]}Hz')
        self.resFreqLabel.pack()
        
        '''
        create the graph button and create the window whwere the graph will be displayed
        we will also create the dropdown menu for the graph's 
        '''
    def showGraphButton(self):
        graphFrame = tk.Frame(self)
        graphFrame.pack()
        
        self.graphDefault = tk.StringVar()
        self.graphDefault.set('Spectogram')
        self.controller.graphButtonClicked(graphFrame)
        menuOptions = self.figures.keys() if self.figures else ["No figures"]
        self.optionDropdown['menu'].delete(0, 'end')
        for option in menuOptions:
            self.optionDropdown['menu'].add_command(label=option, command=tk._setit(self.graphDefault, option, lambda selected=option: self.displayGraph(self.figures[selected], graphFrame) if self.figures else None))
            self.optionDropdown.pack()

        
        
        
    def showStatsButton(self):
        self.controller.StatsButtonClicked()
    
    def hide_message(self):
        self.message_label['text'] = ''
        
    def displayGraph(self, figure,frame):
        """
        Display the graph
        :param figure: the figure that will be displayed (this will be called in the controller)
        :param frame: the frame we wish to assign the graph to
        :return:
        """
        # how we get rid of the old graph when a new one is selected
        if hasattr(self, 'graph'):
            self.graph.get_tk_widget().pack_forget()
        self.graph = FigureCanvasTkAgg(figure, master=frame)
        self.graph.get_tk_widget().pack()
        self.graph.draw()
        pass
    def show_error(self, message):
        """
        Show an error message
        :param message:
        :return:
        """
        self.message_label['text'] = message
        self.message_label['foreground'] = 'red'
        #self.message_label.after(3000, self.hide_message)

    def show_success(self, message):
        """
        Show a message
        :param message:
        :return:
        """
        self.message_label['text'] = message
        self.message_label['foreground'] = 'green'
        #self.message_label.after(3000, self.hide_message)